### PR TITLE
Improvements to the NixOS Hardened Profile

### PR DESCRIPTION
**Pull request by kmcopper**
###### Motivation for this change
Use scudo allocator for the hardened profile by default. Includes changes to compiler-rt to make it more compatible by default with things like man see #65000 for some details. Turning these two options off by default will prevent issues in programs from crashing the malloc, which is critical if we are replacing it as the system malloc. These options are all able to be overridden by simple environment variables even when compiled in so this should be a very safe change. I enabled the allocator zeroing by default in the hardened profile only. The buildSandbox option is currently incompatible with the hardened profile so it should be defaulted to off unless you manually re-enable both it and user namespaces. 

I tested in a VM using the master branch. I used the hardened profile for the guest. I used both standard llvmPackages.compiler_rt and llvmPackages_9.compiler_rt for testing and both work. I mostly tested these options affects on programs by using man but I've also noticed needing these options as an environment variable for qemu to function well.

I feel like this should be backported to 19.09 so that the hardened profile can take advantage of it as was intended in #65000 for 19.09 (albeit with the graphene_malloc allocator). I feel the clang allocator has a much more stable history and should be the mkDefault here. I was unsure if and how to make the environment variable depend on the scudo allocator being the default but I do feel like it's safe to set in to profile anyway since it can be overridden easily.

I wasn't sure if maintainers wanted me to apply the compatibility all the way back to 4 which does still have scudo. I included 8 because I'm not sure what is going to land as the default for 20.03

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @joachifm @lovek323 @dtzWill @7c6f434c

### CODE REVIEW

File: pkgs/development/compilers/llvm/8/compiler-rt.nix [hunk 1]
```diff
@@ -7,6 +7,10 @@ stdenv.mkDerivation {
   nativeBuildInputs = [ cmake python llvm ];
   buildInputs = stdenv.lib.optional stdenv.hostPlatform.isDarwin libcxxabi;
 
+  NIX_CFLAGS_COMPILE = [
+    "-DSCUDO_DEFAULT_OPTIONS=DeleteSizeMismatch=0:DeallocationTypeMismatch=0"
```
Comment (joachifm) [inline pkgs/development/compilers/llvm/8/compiler-rt.nix]: I'd think `cmakeFlags` would be a more natural place to pass this? Or does that not work?

**Comment by joachifm** [comment]
Hi, this LGTM

I think it's inappropriate to backport using scudo by default to 19.09.  Non-essential changes that have some chance of producing observable side-effects violate the implied stable "contract", as I understand it.

If the other changes are deemed safe, we can backport those, however.

The compiler-rt changes should go via staging, due to the darwin mass rebuild.

File: pkgs/development/compilers/llvm/8/compiler-rt.nix (hunk 1 above)
Comment (kmcopper) [reply pkgs/development/compilers/llvm/8/compiler-rt.nix]: I tried there and it ends up uninitialized

**Comment by kmcopper** [comment]
Depends on  #73766 now as per comments.

**Comment by kmcopper** [comment]
I was hoping to get this merged and in 20.03 is it too late now to get this into that?

**Comment by joachifm** [comment]
Sorry I completely forgot about this

File: nixos/modules/profiles/hardened.nix [hunk 2]
```diff
@@ -14,12 +14,17 @@ with lib;
...
+  environment.memoryAllocator.provider = mkDefault "scudo";
+  environment.variables.SCUDO_OPTIONS = mkDefault "ZeroContents=1";
+
   security.hideProcessInformation = mkDefault true;
 
   security.lockKernelModules = mkDefault true;
 
   security.allowUserNamespaces = mkDefault false;
 
+  nix.useSandbox = mkDefault false;
```
Comment (Mic92) [inline nixos/modules/profiles/hardened.nix:26]: Are usernamespaces also not allowed for privileged users?

Comment (joachifm) [reply nixos/modules/profiles/hardened.nix:26]: Debian (used to?) carry a patch that allowed disabling unprivileged ns only.  Vanilla kernel is all or nothing, iirc.

Comment (matthewbauer) [reply nixos/modules/profiles/hardened.nix:26]: Should security.allowUserNamespaces imply nix.useSandbox?

Comment (kmcopper) [reply nixos/modules/profiles/hardened.nix:26]: Here is the relevant debian [patch](https://salsa.debian.org/kernel-team/linux/-/blob/master/debian/patches/debian/add-sysctl-to-disallow-unprivileged-CLONE_NEWUSER-by-default.patch). If allowUserNamespaces is false it should force useSandbox to false as well, if this is a change we want. Ideally I would prefer to rewrite a more scoped patch for the kernel that was a restricted user namespace patch to only allow the build user group and maybe one additional group so namespaces could be used additionally by say a user wanting to benefit from giving chromium/other applications access to the user namespaces but not giving it system wide to every user.

**Comment by Mic92** [comment]
If @joachifm is too busy, maybe @emilazy or @yegortimoshenko can have a look.

**Comment by flokli** [comment]
Urgh - disabling the nix sandbox by default, just because Nix doesn't play well with disabled user namespaces sounds like a very scary hack to me. Not really a fan of 00ac71ab1932b395452209627011a32a63d81897 TBH.

**Comment by joachifm** [comment]
Please elaborate.

**Comment by flokli** [comment]
I consider disabling the Nix sandbox by default dangerous and a huge impurity. Also, see https://github.com/NixOS/nix/pull/3006#pullrequestreview-266714555 why disabling the sandbox alltogether isn't really nice. With this PR merged, shouldn't Nix degrade gracefully?

**Comment by flokli** [comment]
I need to admit, I don't exactly know how and when Nix can enable (some) sandboxing these times, but disabling it alltogether sounds like a bad idea.

**Comment by emilazy** [comment]
I agree, it would be far preferable to merge the kernel patch to allow disabling unprivileged userns creation than it is for the hardened profile to reduce security by silently disabling build sandboxing entirely.

**Comment by dasJ** [comment]
Just found this, I think it would have been better to allow user namespaces rather than disabling sandboxing.
Ever since we switched to the hardened profile, we use `boot.kernel.sysctl."user.max_user_namespaces" = 512;` which sounds a lot less frightening than disabling a sandbox.

**Comment by joachifm** [comment]
Alright, it seems prevailing consensus is to revert that commit. 

IMO, allowing arbitrary code execution on a machine is anathema to "hardening" to begin with. I consider build sandboxing to be mainly a development aide.  I'm unconvinced that disabling the sandbox makes you materially worse off than you already are by handing out access to a C toolchain and all the tools you'd ever need for exploiting bugs in the host kernel. If you're serious about the security of your deployment, do not allow it to run build jobs.

**Comment by emilazy** [comment]
I run the hardened profile on all my NixOS machines and do development on them; I'm sure I'm not alone  The defence in depth hardening features are more useful when running untrusted code, not less. Meanwhile, there's no good reason to use unsandbozef builds at all unless you're in a platform that doesn't support them.

**Comment by flokli** [comment]
Reverted that commit in a8989b353a8266bbd73ce3c2cb91fe4e4dcf7688.

**Comment by emilazy** [comment]
For what it's worth, though I support using a hardened allocator, this currently makes Firefox crash on startup. This doesn't seem to be scudo specific – I get the same results with jemalloc too – but it's unfortunately challenging to work around (seemingly requires a new mount namespace).

It's well-documented that Firefox uses its own jemalloc allocator which can't be overridden without additional effort (see e.g. the [Whonix wiki](https://www.whonix.org/wiki/Hardened_Malloc#cite_note-1)), but a hard SIGSEGV at startup (with the `firefox` package; the binary packages start the crash reporter) seems somewhat unfortunate. Will continue poking, but for now, anyone wanting to use Firefox with the hardened profile can revert to:

```
environment.memoryAllocator.provider = "libc";
```

**Comment by Flakebi** [comment]
Thanks for working on hardening NixOS @kmcopper and @emilazy.
As I was struck by segfaults in the last few days and it took a while to find this issue, I want to add this here:
The freshly enabled allocator leads to segfaults in firefox (as you mentioned), thunderbird, rustc (yeah, all the mozilla related tools) and some background services that get started by evolution.

The easiest way to get a segfault that I found is running `rustc --crate-name ___ --print=file-names --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=cfg /dev/null` (in case it is useful for someone).

**Comment by kmcopper** [comment]
@Flakebi Does `environment.variables.SCUDO_OPTIONS = "" `or `environment.variables.SCUDO_OPTIONS = "ZeroContents=0" ` 

Change the behavior or does it remain? I'm beginning to think made an additional way to scope the  preload especially if we end up switching to graphene in the future as it's much less compatible with various applications than scudo is and I already needed to apply some compatibility fixes to the clang builds in nixpkgs.

**Comment by Flakebi** [comment]
Setting these variables does not change anything.

```bash
env SCUDO_OPTIONS= rustc --crate-name ___ --print=file-names --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=s
ysroot --print=cfg /dev/null
```
 -> Segfault

```bash
env SCUDO_OPTIONS=ZeroContents=0 rustc --crate-name ___ --print=file-names --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=s
ysroot --print=cfg /dev/null
```
 -> Segfault

**Comment by kmcopper** [comment]
Have you tested at all with `environment.memoryAllocator.provider = "graphene-hardened";` on rust as well?

**Comment by Flakebi** [comment]
With `environment.memoryAllocator.provider = "graphene-hardened";`

- rustc works,
- thunderbird works,
- firefox does not work,
- nix stopped working:
```
nixos-rebuild test
building Nix...
building the system configuration...
error: out of memory
```

**Comment by emilazy** [comment]
IIRC the Firefox segfault is because it passes something to libc free that was allocated with jemalloc or vice versa; I don't think there's going to be an easier fix than [implementing the replace-malloc API](https://searchfox.org/mozilla-central/rev/35f2245f17841db50820221f3838ec55744cccd1/memory/build/replace_malloc.h) in a separate library.

**Pull request merged**